src/proof/lrat/lrat_proof.h

```diff
@@ -0,0 +1,207 @@
+
+enum LRATInstructionKind
+{
+  lratDeletion,
+  lratAddition,
+};
+
+struct LRATDeletionData
+{
+  // This idx doesn't really matter, but it's in the format anyway, so we parse
+  // it.
+  ClauseIdx idxOfClause;
+
+  // Clauses to delete
+  std::vector<ClauseIdx> clauses;
+};
+
+// A sequence of locations that will contain unit clauses during unit
+// propegation
+using LRATUPTrace = std::vector<ClauseIdx>;
+
+struct LRATAdditionData
+{
+  // The idx for the new clause
+  ClauseIdx idxOfClause;
+  // The new clause
+  prop::SatClause clause;
+  // UP trace based on the negation of that clause
+  LRATUPTrace atTrace;
+
+  // Clauses that can resolve with `clause` on its first variable,
+  // together with a UP trace after that resolution.
+  // Used for RAT checks.
+  std::vector<std::pair<ClauseIdx, LRATUPTrace>> resolvants;
+};
+
+// This is conceptually an Either<Addition,Deletion>
+struct LRATInstruction
+{
+  LRATInstructionKind kind;
+  LRATAdditionData additionData;
+  LRATDeletionData deletionData;
+};
+
+class LRATProof
+{
+ public:
+
+   /**
+    * @brief Construct an LRAT proof from a DRAT proof, using drat-trim
+    *
+    * @param usedClauses The CNF formula that we're deriving bottom from.
+    *                    It's a map because other parts of the system represent
+    *                    it this way.
+    * @param clauseOrder A record of the order in which those clauses were
+    *                    given to the SAT solver.
+    * @param dratBinary  The DRAT proof from the SAT solver, as a binary stream.
+    */
+  LRATProof(const std::unordered_map<ClauseId, prop::SatClause*>& usedClauses,
+            const std::vector<ClauseId>& clauseOrder,
+            const std::string& dratBinary);
+  /**
+   * @brief Construct an LRAT proof from its textual representation
+   *
+   * @param textualProof the textual encoding of the LRAT proof. See the paper
+   *                     in the file's header comment.
+   */
+  LRATProof(std::istream& textualProof);
+
+  inline const std::vector<LRATInstruction>& getInstructions() const
+  {
+    return d_instructions;
+  }
+
+ private:
+  // The instructions in the proof. Each is a deletion or addition.
+  std::vector<LRATInstruction> d_instructions;
+};
+
+// Prints the literal as a (+) or (-) int
+// Not operator<< b/c that represents negation as ~
+inline std::ostream& textOut(std::ostream& o, const prop::SatLiteral& l)
+{
+  if (l.isNegated())
+  {
+    o << "-";
+  }
+  return o << l.getSatVariable();
+}
+
+// Prints the clause as a space-separated list of ints
+// Not operator<< b/c that represents negation as ~
+inline std::ostream& textOut(std::ostream& o, const prop::SatClause& c)
+{
+  for (const auto l : c)
+  {
+    textOut(o, l) << " ";
+  }
+  return o << "0";
+}
+
+// Prints the trace as a space-separated list of (+) ints with a space at the
+// end.
+inline std::ostream& operator<<(std::ostream& o, const LRATUPTrace& trace)
+{
+  for (const auto i : trace)
```

Comment:
Maybe make this `const auto&` to be consistent with `const auto& idx : del.clauses`:
```suggestion
  for (const auto& i : trace)
```